### PR TITLE
Make fields optional in multi_match query and rely on index.query.default_field by default

### PR DESCRIPTION
**Pull request by jimczi**
This commit adds the ability to send `multi_match` query without providing any `fields`.
When no fields are provided the `multi_match` query will use the fields defined in the index setting `index.query.default_field`
(which in turns defaults to `*`).
The same behavior is already implemented in `query_string` and `simple_query_string` so this change just applies
the heuristic to `multi_match` queries.
Relying on `index.query.default_field` rather than `*` is safer for big mappings that break the 1024 field expansion limit added in 7.0 for all
text queries. For these kind of mappings the admin can change the `index.query.default_field` in order to make sure that exploratory queries using
`multi_match`, `query_string` or `simple_query_string` do not throw an exception.

### CODE REVIEW

File: core/src/main/java/org/elasticsearch/index/query/MultiMatchQueryBuilder.java [hunk 1]
```diff
@@ -793,8 +792,14 @@ protected Query doToQuery(QueryShardContext context) throws IOException {
                 }
             }
         }
-
-        Map<String, Float> newFieldsBoosts = QueryParserHelper.resolveMappingFields(context, fieldsBoosts);
+        Map<String, Float> newFieldsBoosts;
+        if (fieldsBoosts.isEmpty()) {
+            // no fields provided, defaults to index.query.default_field
+            List<String> defaultFields = context.defaultFields();
```
Comment (dakrone) [inline core/src/main/java/org/elasticsearch/index/query/MultiMatchQueryBuilder.java]: Minor nit, we probably don't need to pull this into a variable if we're just immediately passing it into the `parseFieldsAndWeights`

File: core/src/main/java/org/elasticsearch/index/query/MultiMatchQueryBuilder.java [hunk 2]
```diff
@@ -793,8 +792,14 @@ protected Query doToQuery(QueryShardContext context) throws IOException {
...
+            newFieldsBoosts = QueryParserHelper.resolveMappingFields(context, QueryParserHelper.parseFieldsAndWeights(defaultFields));
```
Comment (dakrone) [inline core/src/main/java/org/elasticsearch/index/query/MultiMatchQueryBuilder.java]: Does this need to automatically set `lenient: true` here also?

**Review by dakrone — APPROVED** [review]
LGTM, I left a minor nit and a question

File: core/src/main/java/org/elasticsearch/index/query/MultiMatchQueryBuilder.java (hunk 2 above)
Comment (jimczi) [reply core/src/main/java/org/elasticsearch/index/query/MultiMatchQueryBuilder.java]: Good catch. I pushed another commit to set `lenient: true` if the default field is `*`.

File: core/src/test/java/org/elasticsearch/index/query/MultiMatchQueryBuilderTests.java [hunk 3]
```diff
@@ -338,4 +348,56 @@ public void testToFuzzyQuery() throws Exception {
...
+                new BoostQuery(new TermQuery(new Term(STRING_FIELD_NAME_2, "hello")), 5.0f)
+            ), 0.0f
+        );
+        assertEquals(expected, query);
+
+        context.getIndexSettings().updateIndexMetaData(
+            newIndexMeta("index", context.getIndexSettings().getSettings(), Settings.builder().putList("index.query.default_field",
+                STRING_FIELD_NAME, STRING_FIELD_NAME_2 + "^5", INT_FIELD_NAME).build())
+        );
+        // should fail because lenient defaults to false
```
Comment (dakrone) [inline core/src/test/java/org/elasticsearch/index/query/MultiMatchQueryBuilderTests.java]: Can you add a test here for the `default_field: *` case that it doesn't throw an exception (automatically setting lenient as needed)

**Review by dakrone — APPROVED** [review]
LGTM, I added one more comment about an additional test for the leniency

File: core/src/test/java/org/elasticsearch/index/query/MultiMatchQueryBuilderTests.java (hunk 3 above)
Comment (jimczi) [reply core/src/test/java/org/elasticsearch/index/query/MultiMatchQueryBuilderTests.java]: This is tested by https://github.com/jimczi/elasticsearch/blob/8a130c4e3c6112d889918862221c74d89e153b58/core/src/test/java/org/elasticsearch/index/query/MultiMatchQueryBuilderTests.java#L356

**Pull request merged**